File: mobile/lib/engine/combinatoire.dart
import 'dart:math';

import '../models/horse.dart';

/// Toutes les permutations de longueur [depth] choisies parmi
/// range(0, n), équivalent de itertools.permutations(range(n), depth).
List<List<int>> _permutations(int n, int depth) {
  final result = <List<int>>[];
  if (depth <= 0 || depth > n) return result;

  void recurse(List<int> chosen, List<bool> used) {
    if (chosen.length == depth) {
      result.add(List<int>.from(chosen));
      return;
    }
    for (var x = 0; x < n; x++) {
      if (used[x]) continue;
      used[x] = true;
      chosen.add(x);
      recurse(chosen, used);
      chosen.removeLast();
      used[x] = false;
    }
  }

  recurse(<int>[], List<bool>.filled(n, false));
  return result;
}

/// Probabilité Plackett-Luce de chaque permutation exacte de longueur depth.
/// p_ordre = P_a · (P_b/(1-P_a)) · (P_c/(1-P_a-P_b)) · ... (section 7.12).
Map<List<int>, double> exactPermutationsProbabilities(
    List<double> probabilities, int depth) {
  final result = <List<int>, double>{};
  for (final perm in _permutations(probabilities.length, depth)) {
    var prob = 1.0;
    var remaining = 1.0;
    for (var position = 0; position < perm.length; position++) {
      final p = probabilities[perm[position]];
      if (position == 0) {
        prob *= p;
      } else {
        if (remaining <= 0) {
          prob = 0.0;
          break;
        }
        prob *= p / remaining;
      }
      remaining -= p;
      if (remaining < 0) remaining = 0.0;
    }
    result[perm] = prob;
  }
  return result;
}

class OrdreItem {
  final List<int> ordre;
  final double probabilite;
  OrdreItem(this.ordre, this.probabilite);
}

class EnsembleItem {
  final List<int> ensemble;
  final double probabilite;
  EnsembleItem(this.ensemble, this.probabilite);
}

List<OrdreItem> exactOrdered(List<double> probabilities, int depth) {
  if (depth <= 0) return [];
  final perms = exactPermutationsProbabilities(probabilities, depth);
  final items = perms.entries.map((e) => OrdreItem(e.key, e.value)).toList();
  items.sort((a, b) => b.probabilite.compareTo(a.probabilite));
  return items;
}

List<EnsembleItem> exactUnordered(List<double> probabilities, int depth) {
  final ordered = exactOrdered(probabilities, depth);
  final groups = <String, EnsembleItem>{};
  for (final item in ordered) {
    final sorted = List<int>.from(item.ordre)..sort();
    final key = sorted.join('-');
    final existing = groups[key];
    groups[key] = EnsembleItem(sorted, (existing?.probabilite ?? 0) + item.probabilite);
  }
  final items = groups.values.toList();
  items.sort((a, b) => b.probabilite.compareTo(a.probabilite));
  return items;
}

class PaireItem {
  final int a;
  final int b;
  final double probabilite;
  PaireItem(this.a, this.b, this.probabilite);
}

/// Tirage pondéré sans remise (clés d'Efraimidis-Spirakis), équivalent
/// vectorisé de la version numpy corrigée côté backend : trier par
/// u_i^(1/poids_i) reproduit un tirage séquentiel Plackett-Luce.
Map<String, int> _monteCarloTopIndices(
    List<double> probabilities, int topM, int trials, int seed) {
  final n = probabilities.length;
  final rng = Random(seed);
  final weights = probabilities.map((p) => p < 1e-12 ? 1e-12 : p).toList();
  final counts = <String, int>{};

  for (var t = 0; t < trials; t++) {
    final keys = List<double>.generate(
      n,
      (i) => pow(rng.nextDouble(), 1.0 / weights[i]).toDouble(),
    );
    final order = List<int>.generate(n, (i) => i)
      ..sort((a, b) => keys[b].compareTo(keys[a]));
    final top = order.take(topM).toList()..sort();
    for (var i = 0; i < top.length; i++) {
      for (var j = i + 1; j < top.length; j++) {
        final key = '${top[i]}-${top[j]}';
        counts[key] = (counts[key] ?? 0) + 1;
      }
    }
  }
  return counts;
}

List<PaireItem> monteCarloPlaceProbabilities(
    List<double> probabilities, int topM,
    {int trials = 20000, int seed = 0}) {
  final n = probabilities.length;
  if (n == 0 || topM <= 0) return [];
  final m = min(topM, n);
  final counts = _monteCarloTopIndices(probabilities, m, trials, seed);
  return counts.entries.map((e) {
    final parts = e.key.split('-');
    return PaireItem(int.parse(parts[0]), int.parse(parts[1]), e.value / trials);
  }).toList();
}

class ComboLabel {
  final List<int?> dossards;
  final List<String> noms;
  final double probabilite;
  ComboLabel(this.dossards, this.noms, this.probabilite);
}

class ComboOrdreDesordre {
  final List<ComboLabel> ordre;
  final List<ComboLabel> desordre;
  ComboOrdreDesordre(this.ordre, this.desordre);
}

class Combinaisons {
  final ComboOrdreDesordre couple;
  final ComboOrdreDesordre trio;
  final ComboOrdreDesordre quarte;
  final ComboOrdreDesordre quinte;
  final List<ComboLabel> couplePlace;
  final List<ComboLabel> deuxSurQuatre;

  Combinaisons({
    required this.couple,
    required this.trio,
    required this.quarte,
    required this.quinte,
    required this.couplePlace,
    required this.deuxSurQuatre,
  });
}

ComboLabel _label(List<Horse> reels, List<int> indices) {
  return ComboLabel(
    indices.map((i) => reels[i].numPmu).toList(),
    indices.map((i) => reels[i].nom).toList(),
    0,
  );
}

ComboOrdreDesordre _comboOrdreDesordre(List<Horse> reels, int nbReels,
    List<int> ordreFavoris, int depth, int poolSize) {
  final pool = ordreFavoris.take(min(poolSize, nbReels)).toList();
  if (nbReels < depth || pool.length < depth) {
    return ComboOrdreDesordre([], []);
  }

  final poolProbs = pool.map((i) => reels[i].probabilite).toList();
  final ordered = exactOrdered(poolProbs, depth).take(3);
  final unordered = exactUnordered(poolProbs, depth).take(3);

  final ordreOut = ordered.map((item) {
    final indices = item.ordre.map((localI) => pool[localI]).toList();
    final label = _label(reels, indices);
    return ComboLabel(label.dossards, label.noms, item.probabilite);
  }).toList();

  final desordreOut = unordered.map((item) {
    final indices = item.ensemble.map((localI) => pool[localI]).toList();
    final label = _label(reels, indices);
    return ComboLabel(label.dossards, label.noms, item.probabilite);
  }).toList();

  return ComboOrdreDesordre(ordreOut, desordreOut);
}

List<ComboLabel> _comboPlace(List<Horse> reels, int nbReels,
    List<double> probabilitesChamp, int topM) {
  final counts = monteCarloPlaceProbabilities(probabilitesChamp, topM);
  final reelsPairs = counts.where((p) => p.a < nbReels && p.b < nbReels).toList()
    ..sort((a, b) => b.probabilite.compareTo(a.probabilite));
  return reelsPairs.take(3).map((p) {
    final label = _label(reels, [p.a, p.b]);
    return ComboLabel(label.dossards, label.noms, p.probabilite);
  }).toList();
}

/// Combinaisons calculées sur le résultat complet d'analyseCourse (réels +
/// outsiders virtuels). nbReels est le nombre de chevaux réellement saisis :
/// seuls eux peuvent être nommés dans une combinaison, mais la probabilité de
/// chaque ordre reste calculée sur le champ complet (section 7.12).
Combinaisons buildCombinaisons(List<Horse> horses, int nbReels) {
  final probabilitesChamp = horses.map((h) => h.probabilite).toList();
  final reels = horses.take(nbReels).toList();
  final ordreFavoris = List<int>.generate(nbReels, (i) => i)
    ..sort((a, b) => reels[b].probabilite.compareTo(reels[a].probabilite));

  return Combinaisons(
    couple: _comboOrdreDesordre(reels, nbReels, ordreFavoris, 2, 8),
    trio: _comboOrdreDesordre(reels, nbReels, ordreFavoris, 3, 7),
    quarte: _comboOrdreDesordre(reels, nbReels, ordreFavoris, 4, 7),
    quinte: _comboOrdreDesordre(reels, nbReels, ordreFavoris, 5, 7),
    couplePlace: _comboPlace(reels, nbReels, probabilitesChamp, 3),
    deuxSurQuatre: _comboPlace(reels, nbReels, probabilitesChamp, 4),
  );
}

File: backend/app/engine/combinatoire.py
import itertools
from collections import defaultdict
from typing import Dict, List, Sequence, Tuple

import numpy as np

from app.engine.scoring import HorseAnalysis


def exact_permutations_probabilities(probabilities: Sequence[float], depth: int) -> Dict[Tuple[int, ...], float]:
    result: Dict[Tuple[int, ...], float] = {}
    for perm in itertools.permutations(range(len(probabilities)), depth):
        prob = 1.0
        remaining = 1.0
        for position, index in enumerate(perm):
            p = probabilities[index]
            if position == 0:
                prob *= p
            else:
                if remaining <= 0:
                    prob = 0.0
                    break
                prob *= p / remaining
            remaining -= p
            if remaining < 0:
                remaining = 0.0
        result[perm] = prob
    return result


def exact_ordered(probabilities: Sequence[float], depth: int) -> List[Dict[str, object]]:
    if depth <= 0:
        return []
    perms = exact_permutations_probabilities(probabilities, depth)
    return [
        {"ordre": perm, "probabilite": prob}
        for perm, prob in sorted(perms.items(), key=lambda item: item[1], reverse=True)
    ]


def exact_unordered(probabilities: Sequence[float], depth: int) -> List[Dict[str, object]]:
    ordered = exact_ordered(probabilities, depth)
    groups: Dict[Tuple[int, ...], float] = defaultdict(float)
    for item in ordered:
        comb = tuple(sorted(item["ordre"]))
        groups[comb] += item["probabilite"]
    return [
        {"ensemble": comb, "probabilite": prob}
        for comb, prob in sorted(groups.items(), key=lambda item: item[1], reverse=True)
    ]


def monte_carlo_place_probabilities(probabilities: Sequence[float], top_m: int, trials: int = 20000, seed: int = 0) -> Dict[Tuple[int, int], float]:
    n = len(probabilities)
    if n == 0 or top_m <= 0:
        return {}
    top_m = min(top_m, n)

    # Tirage pondéré sans remise, vectorisé (clés d'Efraimidis-Spirakis :
    # trier par u_i^(1/poids_i) équivaut à un tirage séquentiel Plackett-Luce).
    rng = np.random.default_rng(seed)
    weights = np.clip(np.asarray(probabilities, dtype=float), 1e-12, None)
    keys = rng.random((trials, n)) ** (1.0 / weights)
    top_indices = np.argsort(-keys, axis=1)[:, :top_m]

    counts: Dict[Tuple[int, int], int] = {}
    for row in top_indices:
        for a, b in itertools.combinations(sorted(row.tolist()), 2):
            counts[(a, b)] = counts.get((a, b), 0) + 1
    return {pair: count / trials for pair, count in counts.items()}


def _label(reels: Sequence[HorseAnalysis], indices: Sequence[int]) -> Dict[str, object]:
    return {
        "dossards": [reels[i].num_pmu for i in indices],
        "noms": [reels[i].nom for i in indices],
    }


def _combo_ordre_desordre(reels: Sequence[HorseAnalysis], nb_reels: int, ordre_favoris: List[int], depth: int, pool_size: int) -> Dict[str, List[Dict[str, object]]]:
    pool = ordre_favoris[:min(pool_size, nb_reels)]
    if nb_reels < depth or len(pool) < depth:
        return {"ordre": [], "desordre": []}

    pool_probs = [reels[i].probabilite for i in pool]
    ordered = exact_ordered(pool_probs, depth)[:3]
    unordered = exact_unordered(pool_probs, depth)[:3]
    return {
        "ordre": [
            {**_label(reels, [pool[i] for i in item["ordre"]]), "probabilite": item["probabilite"]}
            for item in ordered
        ],
        "desordre": [
            {**_label(reels, [pool[i] for i in item["ensemble"]]), "probabilite": item["probabilite"]}
            for item in unordered
        ],
    }


def _combo_place(reels: Sequence[HorseAnalysis], nb_reels: int, probabilites_champ: Sequence[float], top_m: int) -> List[Dict[str, object]]:
    counts = monte_carlo_place_probabilities(probabilites_champ, top_m)
    reels_pairs = {pair: p for pair, p in counts.items() if pair[0] < nb_reels and pair[1] < nb_reels}
    best = sorted(reels_pairs.items(), key=lambda item: item[1], reverse=True)[:3]
    return [{**_label(reels, list(pair)), "probabilite": p} for pair, p in best]


def build_combinaisons(horses: Sequence[HorseAnalysis], nb_reels: int) -> Dict[str, object]:
    """Combinaisons calculées sur le résultat complet d'analyse_course (réels +
    outsiders virtuels). nb_reels est le nombre de chevaux réellement saisis :
    seuls eux peuvent être nommés dans une combinaison, mais la probabilité de
    chaque ordre reste calculée sur le champ complet (les outsiders virtuels
    absorbent leur part de probabilité, cf. section 7.12 du cahier des charges)."""
    probabilites_champ = [h.probabilite for h in horses]
    reels = list(horses[:nb_reels])
    ordre_favoris = sorted(range(nb_reels), key=lambda i: reels[i].probabilite, reverse=True)

    return {
        "couple": _combo_ordre_desordre(reels, nb_reels, ordre_favoris, 2, 8),
        "trio": _combo_ordre_desordre(reels, nb_reels, ordre_favoris, 3, 7),
        "quarte": _combo_ordre_desordre(reels, nb_reels, ordre_favoris, 4, 7),
        "quinte": _combo_ordre_desordre(reels, nb_reels, ordre_favoris, 5, 7),
        "couple_place": _combo_place(reels, nb_reels, probabilites_champ, 3),
        "deux_sur_quatre": _combo_place(reels, nb_reels, probabilites_champ, 4),
    }
